✨ 3.0.0 (#13)

CollierCZ opened the pull request:
* Use / extend Svelte types

* Improved validation errors

* Updated eslint and packages

* Load schema with HMR

* Smarter options, and schemaDirectory defaults

* Support variables schema

* Refactored partials loading

* Refactor renderer

* Try Vitest commonmark schema

* Cleanup escapeHtml

* Finished basic test

* Refactored components and import string

* Add component string tests

* Add layout, and only import needed components

* Complete schema options and tests

* Adopt vite normalizePath

* Update deps and exports

* Update README

---------

CollierCZ committed:
✨ Add option for typographer

CollierCZ committed:
✨ Allow comments in Markdoc

CollierCZ committed:
💥 Set default file extension as "official" extension

Official: https://github.com/markdoc/markdoc/discussions/486#discussioncomment-8623962

CollierCZ committed:
📦 Bump major version

For new default with extensions

CollierCZ committed:
📦 Proper built files

CollierCZ committed:
✨ Improve defaults, error handling, and importing (#13)

* Use / extend Svelte types

* Improved validation errors

* Updated eslint and packages

* Load schema with HMR

* Smarter options, and schemaDirectory defaults

* Support variables schema

* Refactored partials loading

* Refactor renderer

* Try Vitest commonmark schema

* Cleanup escapeHtml

* Finished basic test

* Refactored components and import string

* Add component string tests

* Add layout, and only import needed components

* Complete schema options and tests

* Adopt vite normalizePath

* Update deps and exports

* Update README

---------

Co-authored-by: Christopher Hugentobler <hugentoblerc@gmail.com>

CollierCZ committed:
✨ Improve defaults, error handling, and importing (#13)

* Use / extend Svelte types

* Improved validation errors

* Updated eslint and packages

* Load schema with HMR

* Smarter options, and schemaDirectory defaults

* Support variables schema

* Refactored partials loading

* Refactor renderer

* Try Vitest commonmark schema

* Cleanup escapeHtml

* Finished basic test

* Refactored components and import string

* Add component string tests

* Add layout, and only import needed components

* Complete schema options and tests

* Adopt vite normalizePath

* Update deps and exports

* Update README

---------

Co-authored-by: Christopher Hugentobler <hugentoblerc@gmail.com>

hugentobler committed:
✅ Make existing vitests stricter

CollierCZ commented:
@hugentobler, there are a number of breaking changes here:
* renaming the preprocessor
* removing some exported helper functions
* renaming metadata to frontmatter
And probably some more.

The thing with async tags was a need for an await to handle async nodes/tags (Markdoc.transform has to be awaited), which I had in a code comment but realized I had never pushed that comment 🙈 .

But with the breaking changes, will take me more time to process everything.

CollierCZ commented on src/schema.ts:
```diff
@@ -0,0 +1,150 @@
+          moduleFile = potentialPath; // Already absolute and normalized
+          if (moduleFile.includes("/index.")) {
+            log.info(
+              `Remember to include .ts file extension when importing ${configPartName} into ${moduleFile}`,
```
Not sure I understand this – it appears even in .js files and even when the file extension is included.

hugentobler replied:
Thank you for bringing this up. I remember why I added this. It is optional, but it reminded me a more important change.

My fork is now using ES modules! My intention was to fully support both .js and .ts files.

See package.json is now "type": "module"
And tsconfig.json is now:
```
{
  "compilerOptions": {
    "module": "ES2022",
    "moduleResolution": "node",
    "allowImportingTsExtensions": true,
    // ...
  }
}
```

Moving to ES modules means the user has to include file extensions. Actually I think the linter will pick this up. However, I think I included the log just as a friendly reminder. But in the current format it can't be turned off.

CollierCZ replied:
I'm updating the examples in the ReadMe to fit this. I think the reminder every time is a little much.

CollierCZ committed:
📝 Clarify comments and variable name

CollierCZ committed:
📝 Clarify comments and variable name

CollierCZ committed:
🐛 Allow async functions in config

Such as nodes

CollierCZ committed:
🐛 Allow async functions in config

Such as nodes

CollierCZ committed:
💡 Add and clarify comments

CollierCZ committed:
💡 Add and clarify comments

CollierCZ committed:
🎨 Clarify and simplify

CollierCZ committed:
🎨 Clarify and simplify

hugentobler commented:
Nice, thanks let me work on these:

> renaming the preprocessor

We export both the new and old name for now. 

> removing some exported helper functions

Ah that's fair, I didn't think about these.
Do you rmb in what context you use `escapeMarkdocBrackets` for example? The functionality is still there so we can add it back quite easily.

> renaming metadata to frontmatter

Ah the named module export. I don't have a super strong opinion about this, I think I just thought frontmatter would be make it more obvious.

hugentobler commented:
> The thing with async tags was a need for an await to handle async nodes/tags (Markdoc.transform has to be awaited), which I had in a code comment but realized I had never pushed that comment 🙈 .

I'll take a another look this weekend. Need to refresh my memory of your original code. AFAIK this should already work out-of-the-box but I may be wrong. Will try to find your comment.

CollierCZ commented:
> We export both the new and old name for now.

Good point. I think the difference may be strong enough to just jump a major version right away, but we can see.

> Do you rmb in what context you use escapeMarkdocBrackets

I think it was something that's not necessary anymore. Just means I had to update to remove it, but it works without it.

> I don't have a super strong opinion about this, I think I just thought frontmatter would be make it more obvious.

I think frontmatter is probably better. Just it's different, hence the thinking of it as a breaking change.

>  AFAIK this should already work out-of-the-box but I may be wrong. Will try to find your comment.

The comment wasn't there (only in a staged comment I hadn't pushed). When I tried it with an async node, I got blank pages if I didn't add the `await`. 🤷

CollierCZ committed:
💡 Add and clarify comments

CollierCZ committed:
💡 Add and clarify comments

hugentobler commented:
> When I tried it with an async node, I got blank pages if I didn't add the await. 🤷

Isn't that expected behaviour though? I'm confused.

hugentobler committed:
✨ Collect and export headings

CollierCZ commented:
I meant the await here: https://github.com/CollierCZ/markdoc-svelte/pull/14/files#diff-4fab5baaca5c14d2de62d8d2fceef376ddddcc8e9509d86cfa5643f51b89ce3dR151-R154
I had to add it back in.

CollierCZ committed:
📝 Clean up docs

CollierCZ committed:
📝 Clean up docs

CollierCZ commented:
I think what remains is getting the tests to work and removing unnecessary ones. It's a longer-term ambition to have a docs site powered by this running as a separate package, but that doesn't have to be now if the docs seem enough for now.

hugentobler committed:
🔀 Fix conflicts

hugentobler committed:
✅ Complete tests

hugentobler committed:
📝 Add page heading example

hugentobler commented:
Completed tests. And a few small additions. Kept all your changes. 
Can merge this into 'refactor' branch first: https://github.com/CollierCZ/markdoc-svelte/pull/15

hugentobler commented:
Have time for a final review the coming week?

CollierCZ committed:
📝 Add cross link to example

CollierCZ committed:
🧑‍💻 Clearer test commands

CollierCZ committed:
✅ Add test coverage

CollierCZ committed:
🚨 Explicit typing after test of instance

CollierCZ committed:
📝 Correct error level

CollierCZ committed:
✅ Properly test setting validation levels

CollierCZ committed:
🔇 Don't log messages unless explicitly set to that level

CollierCZ committed:
✅ Test validation message properly

Previous test said message should not be included, but tested for it being included.

CollierCZ committed:
⬆️ Update Node version for tests

CollierCZ committed:
🔇 Pass debug messages only if set to debug

CollierCZ committed:
🔥 Remove unnecessary options

CollierCZ committed:
♻️ Standardize tests and reuse where possible

CollierCZ committed:
♻️ Move options tests together

CollierCZ committed:
🔥 Simplify parsing tests

CollierCZ committed:
♻️ Simplify frontmatter and components tests

And move layout to options

CollierCZ committed:
🔀 Merge pull request #16 from tests

✅ Get tests working and comprehensive

CollierCZ committed:
🔀 Merge branch 'hugentobler-main' into headings

CollierCZ committed:
🚨 Fix some warnings

CollierCZ committed:
🐛 Don't resent heading collection on every round

CollierCZ committed:
🐛 Properly export headings

CollierCZ committed:
📝 Clarify ToC docs

CollierCZ committed:
📸 Update snapshots for headings

CollierCZ committed:
✨ Add default IDs for headings

CollierCZ committed:
📝 Clarify docs for index page

CollierCZ committed:
🔥 Unnecessary files

CollierCZ committed:
🔥 Remove unused constants

CollierCZ committed:
🔥 Remove leftover files

CollierCZ committed:
🔀 Merge pull request #17 from headings

✨ Collect and export headings

CollierCZ commented:
Probably is ready except for tagging and releasing. Give it a check, @hugentobler, since I changed the headings function (and refactored the tests). If you see anything, let me know. Otherwise, I'll get to tagging and releasing as 3.0.0

CollierCZ committed:
💥 Remove old name

CollierCZ committed:
✅ Add test for partials in directory

CollierCZ committed:
🎨 Prettify files

CollierCZ committed:
👷 Add lint test

CollierCZ committed:
💚 Fix import for slugify

CollierCZ committed:
👷 Test build process

CollierCZ committed:
🎨 Prettify

CollierCZ committed:
👷 Fix name of lint step in CI

hugentobler commented:
Thanks a lot. All looks great. I spotted 2 small changes I'd like to make. Can you do it manually as I screwed up my original fork and now can't open a new one.

/src/types.ts
```
_should be_ 
import type { Heading } from "./headings.ts";
_not_
import type { Heading } from "./headings";
```

/src/index.ts
```
export { markdocPreprocess } from "./main.ts";
export type { MarkdocModule } from "./types.ts"; # Please add this line

export { default as Markdoc } from "@markdoc/markdoc";
export type { Config } from "@markdoc/markdoc";
```
It's a super helpful detail for Svelte. If user adds the type when importing, every .svelte page with load function gets the types automatically. E.g. user can write the following and it's typesafe / has typehints.
```
  const { frontmatter, headings } = data.markdown; # The PageProps
  const filteredHeadings = headings?.filter((heading) => heading.level == 2) ?? []; # Heading shape is known
```

CollierCZ committed:
🏷️ Proper type import/export

CollierCZ commented:
Sure thing. Those should both be there now.

hugentobler reviewed (approved):
Nice thanks

CollierCZ committed:
🔖 Version 3.0
